gitpod-io/gitpod
# [server, proxy] Dev: Inject dashboard code at runtime

## Description

This PR introduces a new development mode which is a basic version of what @selfcontained suggested last week. Thanks for sharing that! :pray: 
It introduces a switch at the proxy level (header: `X-Frontend-Dev-URL`) that allows to redirect to arbitrary sources to pull the frontend assets from.

**Example**: Say you want to test a frontend-only change ASAP without waiting for a preview env, but ideally you need some existing test data in that installation.
You go to your browser's header editor of choice and configure the header `X-Frontend-Dev-URL` to point to your workspace port serving that new frontend code, e.g. `https://3000-gitpodio-gitpod-799jhau7k01.ws-eu82.gitpod.io`, and visit gitpod.io. Now you can start play around with that new UI! :sparkles: 

@selfcontained suggested a version which uses a simple query parameter to do the same from frontend, which allows you to send around versions of your new UI and share it with colleagues. That was safe because it would only pull assets from a pre-defined domain/base-path with restricted access. Sadly this does not work if you want to use workspaces as source (incl. hot reload), but I thought this would be cool anyway. :slightly_smiling_face: 

Loom [Demo](https://www.loom.com/share/e52da5a997d84e55ac66231dea4bdd4e):
<a href="https://www.loom.com/share/e52da5a997d84e55ac66231dea4bdd4e">
    <p>Demo</p>
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/e52da5a997d84e55ac66231dea4bdd4e-with-play.gif">
  </a>

FAQ:
- Why proxy level? Because this is basically XSS, and we need to guard against arbitrary changes. Plus we do not have build time variance to disable it for prod only, for instance, so we do it at runtime.
- Speaking of security: Currently, this PR requires you to be a logged in Gitpod user with a permission `developer`. I've no hard feeling here if we want to change that mechanism. Also, we might want to add a "second line of defense" by having an env var on the `proxy` that defaults to off, for instance.

## Related Issue(s)

Fixes #

## How to test
### Prerequisites
- start a workspace ([on this branch](https://gitpod.io/#https://github.com/gitpod-io/gitpod/pull/15725))
 - `cd components/dashboard && yarn start`
 - make port 3000 public
 - sign up at https://gpl-fe-dev.preview.gitpod-dev.com/workspaces

### Test Authorization
- use a browser extension to set header `X-Frontend-Dev-URL` to that ports URL
- visit https://gpl-fe-dev.preview.gitpod-dev.com/workspaces
- note the `Unauthorized` message :heavy_check_mark: 

### Test Forwarding
- get your user the `permission: developer` set (I can help there)
- re-visit https://gpl-fe-dev.preview.gitpod-dev.com/workspaces
- should work and show identical UI (plus react button) :heavy_check_mark: 
- make a change in your local dashboard (change `index.css` for instance)
- re-load https://gpl-fe-dev.preview.gitpod-dev.com/workspaces
- see your changes on the dashboard :tada: 

## Release Notes

```release-note
NONE
```

## Documentation


## Werft options:

      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
